From 99bc33d692814ade8f61795445761a8cb359883b Mon Sep 17 00:00:00 2001
From: Subrata Chatterjee <schatterjee@habana.ai>
Date: Tue, 19 Jul 2022 12:18:44 +0300
Subject: [PATCH] [SW-95811] gaudi2: security: enable jtag

If JTAG_DISABLE (BIT12, offset 58) in efuse is '1' then
ARCBOOT disables JTAG. But the requirement is to
ignore efuse indication of 'JTAG DISABLE' if CPLD RMA enabled.

Following are few scenarios to be handled (in both regular
and recovery flows)

1.rma disable - security enable - JTAG disable
2.rma disable - security disable - JTAG enable
3.rma enable - security disable/enable - JTAG enable

Change done in common security function to handle both normal
and erom recovery flows.

security_enable_jtag_if() is used from arcboot rom.

Signed-off-by: Subrata Chatterjee <schatterjee@habana.ai>
Change-Id: I3ae0028d1c4ff6b0904196b16e24c3251db65645
---

diff --git a/arch/gaudi2/arc/common.c b/arch/gaudi2/arc/common.c
index 98f6c87..7498cd6 100644
--- a/arch/gaudi2/arc/common.c
+++ b/arch/gaudi2/arc/common.c
@@ -1041,3 +1041,8 @@
 
 	return preboot_fake_security_is_sig_validation_enabled();
 }
+
+void ppboot_secure_enable_jtag_if(void)
+{
+	security_enable_jtag_if();
+}
diff --git a/arch/gaudi2/common/security.c b/arch/gaudi2/common/security.c
index 6a266e6..c915a64 100644
--- a/arch/gaudi2/common/security.c
+++ b/arch/gaudi2/common/security.c
@@ -1205,13 +1205,18 @@
 	regmap_writel(reg, mmDCORE0_RTR2_TIF_LBW_USR_RTR);
 }
 
+__weak void ppboot_secure_enable_jtag_if(void) {}
+
 void ppboot_security_init(void)
 {
 	ppboot_secure_elbi();
 	ppboot_secure_hbm();
 
-	if (!preboot_security_is_enabled())
+	if (!preboot_security_is_enabled()) {
+		/* Enable JTAG to override disablement in arcboot */
+		ppboot_secure_enable_jtag_if();
 		return;
+	}
 
 	ppboot_secure_lbw();
 	ppboot_secure_hbw();
